dm/pkg/schema/tracker.go

```diff
@@ -329,3 +359,225 @@ func (tr *Tracker) CreateTableIfNotExists(table *filter.Table, ti *model.TableIn
+					}
+				}
+			}
+		} else {
+			delete(tr.dsTracker.tableInfos, tableID)
+			tctx.Logger.Info("Remove downstream schema tracker", zap.String("tableID", tableID))
+		}
+	}
+}
+
+// getTableInfoByCreateStmt get downstream tableInfo by "SHOW CREATE TABLE" stmt.
+func (tr *Tracker) getTableInfoByCreateStmt(tctx *tcontext.Context, tableID string) (*model.TableInfo, error) {
+	if tr.dsTracker.stmtParser == nil {
+		err := tr.initDownStreamSQLModeAndParser(tctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	createStr, err := utils.GetTableCreateSql(tctx.Ctx, tr.dsTracker.downstreamConn.BaseConn.DBConn, tableID)
+	if err != nil {
+		return nil, dterror.ErrSchemaTrackerCannotFetchDownstreamCreateTableStmt.Delegate(err, tableID)
+	}
+
+	tctx.Logger.Info("Show create table info", zap.String("tableID", tableID), zap.String("create string", createStr))
+	// parse create table stmt.
+	stmtNode, err := tr.dsTracker.stmtParser.ParseOneStmt(createStr, "", "")
+	if err != nil {
+		return nil, dterror.ErrSchemaTrackerInvalidCreateTableStmt.Delegate(err, createStr)
+	}
+
+	ti, err := ddl.MockTableInfo(mock.NewContext(), stmtNode.(*ast.CreateTableStmt), mockTableID)
+	if err != nil {
+		return nil, dterror.ErrSchemaTrackerCannotMockDownstreamTable.Delegate(err, createStr)
+	}
+	return ti, nil
+}
+
+// initDownStreamTrackerParser init downstream tracker parser by default sql_mode.
+func (tr *Tracker) initDownStreamSQLModeAndParser(tctx *tcontext.Context) error {
+	setSQLMode := fmt.Sprintf("SET SESSION SQL_MODE = '%s'", mysql.DefaultSQLMode)
+	_, err := tr.dsTracker.downstreamConn.ExecuteSQL(tctx, []string{setSQLMode})
+	if err != nil {
+		return dterror.ErrSchemaTrackerCannotSetDownstreamSQLMode.Delegate(err, mysql.DefaultSQLMode)
+	}
+	stmtParser, err := utils.GetParserFromSQLModeStr(mysql.DefaultSQLMode)
+	if err != nil {
+		return dterror.ErrSchemaTrackerCannotInitDownstreamParser.Delegate(err, mysql.DefaultSQLMode)
+	}
+	tr.dsTracker.stmtParser = stmtParser
+	return nil
+}
+
+// getDownStreamTi constructs downstreamTable index cache by tableinfo.
+func getDownStreamTi(ti *model.TableInfo, originTi *model.TableInfo) *downstreamTableInfo {
+	var (
+		indexCache       *model.IndexInfo
+		availableUKCache = make([]*model.IndexInfo, 0, len(ti.Indices))
+		hasPk            = false
+	)
+
+	// func for check not null constraint
+	fn := func(i int) bool {
+		return mysql.HasNotNullFlag(ti.Columns[i].Flag)
+	}
+
+	for _, idx := range ti.Indices {
+		if !idx.Primary && !idx.Unique {
+			continue
+		}
+		indexRedict := redirectIndexKeys(idx, originTi)
+		if indexRedict == nil {
+			continue
+		}
+		if idx.Primary {
+			indexCache = indexRedict
+			hasPk = true
+		} else if idx.Unique {
+			// second check not null unique key
+			if !hasPk && isSpecifiedIndexColumn(idx, fn) {
+				indexCache = indexRedict
+			} else {
+				availableUKCache = append(availableUKCache, indexRedict)
+			}
+		}
+	}
+
+	// handle pk exceptional case.
```

Comment:
since PK (even this special PK) is better than UK, we should move this setting `indexCache` before UK's setting

Reply:
yes, this special PK has Higher priority and it is controled by `hasPk`